README
Mojo-Cloudstack

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Mojo::Cloudstack

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Mojo-Cloudstack

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Mojo-Cloudstack

    CPAN Ratings
        http://cpanratings.perl.org/d/Mojo-Cloudstack

    Search CPAN
        http://search.cpan.org/dist/Mojo-Cloudstack/


LICENSE AND COPYRIGHT

Copyright (C) 2015 Holger Rupprecht

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Mojo-Cloudstack

0.01    2015-04-16
        Initial Release.

0.02    2015-05-07
        added dependencies to Build.pl
        added api cache file
        implements an api cache

0.03    2015-05-26
        improved api cache
        implements all responsetypes

0.04    2015-05-26
        cleanups
        removed boilerplate stuff

0.05    2015-05-30
        implements login/logout api calls
        use import and attr from Mojo::Base'

0.06    2015-09-28
        use post request due to CLOUDSTACK-8505 and commit 1c81b241e7914b24b06c3b7b3ee98bc0d3b4f68b only POST methods
        are allowed.

0.07    2015-09-298
        dump on error
        added dep for IO::Socket::SSL


Build.PL
use 5.10.1;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
  module_name        => 'Mojo::Cloudstack',
  license            => 'artistic_2',
  dist_abstract      => 'API Request and class generator for Apache Cloudstack',
  dist_author        => q{Holger Rupprecht <holger.rupprecht@gmx.de>},
  dist_version_from  => 'lib/Mojo/Cloudstack.pm',
  release_status     => 'stable',
  configure_requires => {
    'Module::Build' => 0,
  },
  build_requires => {
    'Test::CPAN::Changes' => 0,
    'Test::More'          => 0,
    'Test::Exception'     => 0,
    'Test::MockObject'    => 0,
  },
  requires => {
    'Mojolicious'       => 0,
    'URL::Encode'       => 0,
    'Digest::HMAC_SHA1' => '1.03',
    'URI::Encode'       => 0,
    'File::HomeDir'     => 0,
  },
  add_to_cleanup     => ['Mojo-Cloudstack-*'],
  create_makefile_pl => 'traditional',
);

$builder->create_build_script();


ex/createvm.pl
#!/usr/bin/perl

use Mojo::Util 'slurp';
use Data::Dumper 'Dumper';
use Mojo::Cloudstack;

my $api_key = slurp("/home/holger/.mojo_cloudstack/api_key");
chomp $api_key;
my $secret_key = slurp("/home/holger/.mojo_cloudstack/secret_key");
chomp $secret_key;

my $cs = Mojo::Cloudstack->new(
  host       => "172.29.0.10",
  path       => "/client/api",
  port       => "443",
  scheme     => "https",
  api_key    => $api_key,
  secret_key => $secret_key,
);

my $project_id = $cs->listProjects(
  name => 'Blueprint-Customer'
)->[0]->id;

my $zones = $cs->listZones;
my $zone1 = $zones->first;

my $sos = $cs->listServiceOfferings(id => "44741d1a-242d-43dc-8dad-82af1d549866");
my $so = $sos->first;

my $templates = $cs->listTemplates(
  templatefilter => 'featured',
  id => 'be8251d3-4d64-47fe-8ab2-4d77c0d6b068'
);
my $t = $templates->first;


#warn Dumper $zone1, $so, $templates;
#warn Dumper $zone1->name, $so->name, $t->name ;

my $vmreq = $cs->deployVirtualMachine(
  serviceofferingid => $so->id,
  templateid => $t->id,
  zoneid => $zone1->id,
  projectid => $project_id,
);

#warn Dumper $vmreq;

my $jobid = $vmreq->jobid;
warn "JOBID $jobid";

my $jobstatus = 0;
my $res;
$| = 1;

while(not $jobstatus == 1){
  $res = $cs->queryAsyncJobResult(
    jobid => $jobid
  );
  $jobstatus = $res->jobstatus;
  print '.';
  sleep 1;
}

print "\n";
print "vm_name  : " . $res->jobresult->name     . "\n";
print "password : " . $res->jobresult->password . "\n";

#warn Dumper $res;



lib/Mojo/Cloudstack/Base.pm
package Mojo::Cloudstack::Base;
use Mojo::Base -base;

sub new {
  my $class = shift;
  my $cs_class = shift;
  my $self = bless shift, $cs_class;

  no strict 'refs';
  no warnings 'redefine';
  foreach my $key (keys %$self){
    #TODO test $self->attr($key => sub { $self->{$key} });
    #TODO disable no warnings 'redefine';
    *{"${cs_class}::${key}"} = sub {
      my ($self, $value) = @_;
      return $value ? (($self->{$key} = $value) and $self) : $self->{$key};
    };
    if($key eq 'jobresult'){
      my $otype = (keys %{$self->{jobresult}})[0];
      $self->$key(
        Mojo::Cloudstack::Base->new('Mojo::Cloudstack::' . ucfirst($otype), $self->{$key}{$otype})
      );
    }
  }
  return $self;
}

1;

__END__

=pod

=head1 NAME

Mojo::Cloudstack::Base

=head1 DESCRIPTION

This is the base class for all autogenerated Cloudstack API call response classes

=head1 SYNOPSIS

  use Mojo::Cloudstack::Base;

=head1 METHODS

=head2 new

  my $class = Mojo::Cloudstack::Base->new('Mojo::Cloudstack::Virualmachine', $memberhash);

=cut


lib/Mojo/Cloudstack.pm
package Mojo::Cloudstack;

use Mojo::Base 'Mojo::UserAgent';
use Mojo::Parameters;
use Mojo::URL;
use Mojo::UserAgent;
use Mojo::JSON 'j';
use Mojo::Util 'slurp';
use Mojo::Collection 'c';
use Mojo::Cloudstack::Base;
use Mojo::Cloudstack::Api;
use Digest::HMAC_SHA1 qw(hmac_sha1 hmac_sha1_hex);
use MIME::Base64;
use URI::Encode 'uri_encode';
use File::HomeDir;
use Data::Dumper 'Dumper';

has 'host'        => "localhost";
has 'path'        => "/client/api";
has 'port'        => "8080";
has 'scheme'      => "https";
has 'api_key'     => "";
has 'secret_key'  => "";
has 'responsetypes' => '';
has 'api_cache'   => sub {
  my $self = shift;
  $self->_load_api_cache;
  $self->__build_responsetypes;
};

our $VERSION = '0.06';
our $AUTOLOAD;

chomp(our $user = `whoami`);
our $cf = File::HomeDir->users_home($user) . "/.cloudmojo/api.json";

sub _build_request {
  my ($self, $params) = @_;
  my $baseurl = sprintf ("%s://%s:%s%s?", $self->scheme, $self->host, $self->port, $self->path);
  $params->{ apiKey }   = $self->api_key;
  $params->{ response } = 'json';
  my $secret_key = $self->secret_key;

  my $req_params = Mojo::Parameters->new();
  foreach my $p (sort keys %$params) {
    $req_params->param($p => uri_encode($params->{ $p }));
  }
  my $params_str = lc ($req_params->to_string);
  my $digest = hmac_sha1($params_str, $secret_key);
  my $base64_encoded = encode_base64($digest);
  chomp ($base64_encoded);

  my $uri_encoded = uri_encode($base64_encoded,1);
  my $url = Mojo::URL->new($baseurl);
  $url->query($req_params->to_string);

  return $url->to_string . '&signature='.$uri_encoded;

}

sub AUTOLOAD {
  my $self = shift;
  (my $command = $AUTOLOAD) =~ s/.*:://;
  my %params = @_;
  $params{command} = $command;
  my $req = $self->_build_request(\%params);
  my $res = $self->post($req)->res;
  my $items = $res->json;
  #warn Dumper 'ITEMS', $items;
  die sprintf("Could not get response for %s %s %s", $req,  $res->code, $res->message) unless $items;
  my $responsetype = (keys %$items)[0];

  if($responsetype =~ /^(login|activate|add|archive|assign|associate|attach|authorize|change|configure|copy|create|delete|deploy|destroy|detach|disable|disassociate|enable|error|expunge|extract|get|list|lock|migrate|query|reboot|recover|register|remove|replace|reset|resize|restart|restore|revert|revoke|scale|start|stop|suspend|update|upload)(.*)(response)$/){
    my ($otype, $oname, $oresponse) = ($1, $2, $3);
    $items->{$responsetype}{_cs} = $self unless $oname eq 'apis';
    if($oname eq 'apis'){
      $self->__write_api_cache($items);
      return $items;
    } elsif($otype eq 'list'){
      if($oname =~ /(s)$/){
        $oname =~ s/$1$//;
      }
      return c(
        map {
          $_->{_cs} = $self;
          Mojo::Cloudstack::Base->new('Mojo::Cloudstack::' . ucfirst($oname),$_);
        } @{$items->{$responsetype}{$oname}}
      );
    } elsif($otype eq 'query'){
      return  Mojo::Cloudstack::Base->new('Mojo::Cloudstack::AsyncJobResult', $items->{$responsetype});
    } elsif(exists $items->{$responsetype}{errorcode}){
      return  Mojo::Cloudstack::Base->new('Mojo::Cloudstack::Error', $items->{$responsetype});
    } elsif($otype =~ /^log(in|out)$/){
      return  Mojo::Cloudstack::Base->new('Mojo::Cloudstack::' . ucfirst($&), $items->{$responsetype});
    } else {
      return  Mojo::Cloudstack::Base->new('Mojo::Cloudstack::' . (exists $items->{$responsetype}{jobid}
        ? 'AsyncJobRequest'
        : ucfirst($oname)), $items->{$responsetype});
    }
  } else {
    die "unknown response type $responsetype for reqest \n$req";
  }

}

sub sync {
  return shift->_load_api_cache(1);
}

sub _load_api_cache {
  my ($self, $force) = @_;
  $self->api_cache($self->__build_api_json($force))
    and return $self->api_cache
      if ($force or (not -f $cf));
  #TODO File::ShareDir

  $self->api_cache(j(slurp $cf)) if -f $cf;
  return $self->api_cache;
}

sub __build_responsetypes {
  my ($self) = @_;
  my $apis = $self->api_cache;
  my %responsetypes = map { (split(/[A-Z]/,$_->{name},2))[0] => 1  }
    @{ $apis->{listapisresponse}{api} };
  $responsetypes{error} = 1;
  $self->responsetypes(join('|', sort keys %responsetypes));
  die $self->responsetypes;
}

sub __build_api_json {
  my ($self, $force) = @_;
  my $apis = $self->listApis;
  $self->__write_api_cache($apis, $force);
  return $apis;
}

sub __write_api_cache {
  my($self, $apis, $force) = @_;
  my $cachedir = File::HomeDir->users_home($user) . "/.cloudmojo";
  mkdir $cachedir unless -d $cachedir;
  my $cachefile = "$cachedir/api.json";
  unlink $cachefile if $force;
  unless (-f $cachefile){
    open my $cf, ">$cachefile";
    print $cf j($apis);
    close $cf;
  }
  return $apis;

}

1;

__END__

=pod

=head1 NAME

Mojo::Cloudstack

=head1 DESCRIPTION

Generic Class for Cloudstack API calls
The goal is to have blessed classes generated by AUTOLOAD, which are replaced
step by step by 'real' classes when enhanced functionality is needed.


=head1 SYNOPSIS

  use Mojo::Cloudstack;

=head1 METHODS

=head2 new

  my $cs = Mojo::Cloudstack->new(
    host       => "cloustack.local",
    path       => "/client/api",
    port       => "443",
    scheme     => "https",
    api_key    => $api_key,
    secret_key => $secret_key,
  );

  my $vmreq = $cs->deployVirtualMachine(
    serviceofferingid => $so->id,
    templateid => $t->id,
    zoneid => $zone1->id,
    projectid => $project_id,
  );

=head2 sync

Sync possible API calls against Cloudstack with your account
dependend on your permisions.

  $cs->sync;

=head2 _build_request

  my $params = Mojo::Parameters->new('command=listUsers&response=json');
  $params->append(apiKey => $cs->api_key);
  my $req = $cs->_build_request($params->to_hash)

=head2 api_key

API Key from Cloudstack

=head2  host

Host to connect to

=head2 path

URL path to API

=head2 port

Port to connect to

=head2 scheme

URL scheme http | https

=head2 secret_key

Secret Key from Cloudstack

=cut
